rfcs/DeferStream.md

```diff
@@ -17,4 +17,110 @@ A `@defer` directive would allow clients to annotate fragments that are non-crit
+
+These solutions require the GraphQL client to efficiently process multiple responses at the same time. (Relay support added here: https://github.com/facebook/relay/commit/b4c92a23ae061943ea7a2ddb5e2f7686d3af8c0e)
+
+1. __Client relies on transport to receive multiple responses.__ If the incremental responses are being sent over HTTP connection with chunked encoding, the client may receive multiple responses in a single read of HTTP stream and process them at the same time. This is only likely to happen when the responses are small and sent very close together. This would not work for all possible transport types, e.g. web sockets where each frame is received separately.
+
+For example, the client might receive several responses at once:
+```
+
+---
+Content-Type: application/json
+Content-Length: 125
+
+{
+    "path": ["viewer","itemSearch","edges",5],
+    "data": {
+        "node": {
+            "item": {
+                "attribute": "Vintage 1950s Swedish Scandinavian Modern"
+            }
+        }
+    }
+}
+
+---
+Content-Type: application/json
+Content-Length: 126
+
+{
+    "path": ["viewer","itemSearch","edges",6],
+    "data": {
+        "node": {
+            "item": {
+                "attribute": "Mid-20th Century Italian Hollywood Regency"
+            }
+        }
+    }
+}
+
+---
+Content-Type: application/json
+Content-Length: 124
+
+{
+    "path": ["viewer","itemSearch","edges",7],
+    "data": {
+        "node": {
+            "item": {
+                "attribute": "Vintage 1950s Italian Mid-Century Modern"
+            }
+        }
+    }
+}
+```
+
+It could then process each of these before triggering a re-render.
+
+2. __Client side debounce.__ GraphQL clients can debounce the processing of responses before triggering a re-render. For a query that contains `@defer` or `@stream`, the client will wait a predetermined amount of time starting from when a response is received. If any additional responses are received in that time, it can process the results in one batch. This has the downside of adding latency; if no additional responses are receiving in the timeout period, the processing of the initial response is delayed by the length of the debounce timeout.
+3. __Server sends batched responses.__ This approach changes the spec to allow GraphQL to return either the current GraphQL Response map, or a list of GraphQL Response maps. This gives the server the flexibility to determine when it is beneficial to group incremental responses together. If several responses are ready at the same time, the server can deliver them together. The server may also have knowledge of how long resolvers will take to resolve and could choose to debounce.
```

Comment:
My intuition is that this would be the direction we want to pursue precisely for the flexibility you mention here. When experimenting at Facebook, we tried client side debouncing first, and due to the complexity of figuring out a reasonable interval, we ended up with a naive debouncing which is always rendering twice, once for the initial payload and once after all incremental payloads have arrived. 
That turns out to be not granular enough, so we implemented server side batching. One complication is that the GraphQL executor actually doesn't have a good sense of how expensive a resolver is. Therefore, it's hard for the executor to make a decision on the fly while executing a query. It's more applicable if we outsource the decision of how to batch to the underlying @stream and @defer resolvers. In the end, we only implemented batching for @stream since a global batching at the query level means @deferred and @streamed resolvers need to communicate with each other in some sense. 

I wonder if we can offer practical suggestions for servers to implement server side batching. If not, should we instead propose something like @stream level batching?